minidump-processor/src/symbols.rs

```diff
@@ -350,3 +350,251 @@ pub fn simple_symbol_supplier(symbol_paths: Vec<PathBuf>) -> impl SymbolSupplier
 pub fn string_symbol_supplier(modules: HashMap<String, String>) -> impl SymbolSupplier {
     breakpad_symbols::StringSymbolSupplier::new(modules)
 }
+
+pub mod debuginfo {
+    use super::*;
+    use breakpad_symbols::SymbolFile;
+    use cachemap::CacheMap;
+    use futures_util::lock::Mutex;
+    use memmap2::Mmap;
+    use std::collections::HashMap;
+    use std::fs::File;
+    use std::path::Path;
+    use symbolic::{
+        cfi::CfiCache,
+        debuginfo::{self, Object},
+    };
+
+    /// A symbol provider which gets symbol information from the crashing binaries on the local
+    /// system.
+    #[derive(Default)]
+    pub struct DebugInfoSymbolProvider {
+        /// If a file fails to load for any reason, None is stored.
+        loaded: CacheMap<PathBuf, Lazy<Option<DebugInfo>>>,
+    }
+
+    #[derive(Default)]
+    struct Lazy<T>(Mutex<Option<T>>);
+
+    impl<T> Lazy<T> {
+        pub async fn get<F: FnOnce() -> T>(&self, if_missing: F) -> &T {
+            let mut guard = self.0.lock().await;
+            if guard.is_none() {
+                *guard = Some(if_missing());
+            }
+            debug_assert!(guard.is_some());
+            // # Safety
+            // The inner value is guaranteed to be set, and it will never be changed again so we
+            // may return a &T tied to the lifetime of &self.
+            unsafe {
+                (guard.as_ref().unwrap_unchecked() as *const T)
+                    .as_ref()
+                    .unwrap_unchecked()
+            }
+        }
+    }
+
+    struct DebugInfo {
+        functions: Vec<Function>,
+        unwind_symbol_file: Option<SymbolFile>,
+    }
+
+    impl DebugInfo {
+        pub fn new(file: &Path) -> Option<Self> {
+            let file = File::open(file).ok()?;
+            // # Safety
+            // The file is presumably read-only (being some binary or debug info file).
+            let mapped = unsafe { Mmap::map(&file) }.ok()?;
+
+            let object = debuginfo::Object::parse(&mapped).ok()?;
+            Some(Self::from_object(object))
+        }
+
+        pub fn from_object(object: Object) -> Self {
+            let mut functions: Vec<Function> = object
+                .debug_session()
+                .ok()
+                .map(|session| {
+                    session
+                        .functions()
+                        .filter_map(Result::ok)
+                        .map(Into::into)
+                        .collect()
+                })
+                .unwrap_or_default();
+            functions.sort_unstable_by_key(|f| f.address);
+
+            let unwind_symbol_file = CfiCache::from_object(&object)
+                .ok()
+                .and_then(|cache| SymbolFile::from_bytes(cache.as_slice()).ok());
```

Comment:
I think we had this discussion with Sentry folks a while back. The `symbolic` model is optimized for server-side processing, so building symcache (and presumably cficache) objects is handled as an expensive up-front operation to speed up future access. Most debug symbol formats are, in fact, optimized for random lookups since that's the primary use-case for debuggers. The unwind tables (such as CFI), in particular are *highly* optimized for random access and runtime performance, since they are designed for unwinding exceptions. @mstange is probably the best person to speak to these concerns.

Reply:
But to answer the original question, I just made a test and ran on nightly libxul.
Over 10 runs:
> Time (mean ± σ):      1.437 s ±  0.025 s    [User: 1.338 s, System: 0.096 s]

Reply:
That's fine for a first stab at this, thanks! Naturally `samply-symbols`/`wholesym` will be preferable as soon as we can use them, then we won't have to pay the upfront costs like @luser mentioned.